src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -0,0 +1,1851 @@
+                    if (_consumed >= (uint)localBuffer.Length)
+                    {
+                        return true;
+                    }
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.AllowComments)
+                    {
+                        // This is necessary to avoid throwing when the user has 1 or more comments as the first token
+                        // OR if there is a comment after a single, non-primitive value.
+                        // In this mode, ConsumeValue consumes the comment and we need to return it as a token.
+                        // along with future comments in subsequeunt reads.
+                        if (_tokenType == JsonTokenType.Comment || localBuffer[_consumed] == JsonConstants.Slash)
+                        {
+                            return true;
+                        }
+                    }
+                    else
+                    {
+                        Debug.Assert(_readerOptions.CommentHandling == JsonCommentHandling.SkipComments);
+                        if (_tokenType == JsonTokenType.StartObject || _tokenType == JsonTokenType.StartArray)
+                        {
+                            _isNotPrimitive = true;
+                        }
+                        return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, localBuffer[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpace()
+        {
+            // Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localBuffer = _buffer;
+            for (; _consumed < localBuffer.Length; _consumed++)
+            {
+                byte val = localBuffer[_consumed];
+
+                // JSON RFC 8259 section 2 says only these 4 characters count, not all of the Unicode defintions of whitespace.
+                if (val != JsonConstants.Space &&
+                    val != JsonConstants.CarriageReturn &&
+                    val != JsonConstants.LineFeed &&
+                    val != JsonConstants.Tab)
+                {
+                    break;
+                }
+
+                if (val == JsonConstants.LineFeed)
+                {
+                    _lineNumber++;
+                    _bytePositionInLine = 0;
+                }
+                else
+                {
+                    _bytePositionInLine++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method contains the logic for processing the next value token and determining
+        /// what type of data it is.
+        /// </summary>
+        private bool ConsumeValue(byte marker)
+        {
+        // Using goto to avoid recursive calls.
+        Begin:
```

Comment:
And that can't be accomplished with a `while (true)` and `continue` instead of `goto`?